controlplane/kubeadm/internal/cluster_test.go

```diff
@@ -178,6 +154,178 @@ func TestGetMachinesForCluster(t *testing.T) {
+	ctx := context.Background()
+	// create an envtest for all subtests
+	testEnv := helpers.NewTestEnvironment()
+	defer func() {
+		if testEnv != nil {
+			g.Expect(testEnv.Stop()).To(Succeed(), "stopping envtest")
+		}
+	}()
+	go func() {
+		g.Expect(testEnv.StartManager()).To(Succeed())
+	}()
+
+	// Create an etcd secret with valid certs
+	key, err := certs.NewPrivateKey()
+	g.Expect(err).ToNot(HaveOccurred())
+	cert, err := getTestCACert(key)
+	g.Expect(err).ToNot(HaveOccurred())
+	etcdSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "my-cluster-etcd",
+			Namespace: "my-namespace",
+		},
+		Data: map[string][]byte{
+			secret.TLSCrtDataName: certs.EncodeCertPEM(cert),
+			secret.TLSKeyDataName: certs.EncodePrivateKeyPEM(key),
+		},
+	}
+
+	// Create kubeconfig secret
+	// Store the envtest config as the contents of the kubeconfig secret.
+	// This way we are using the envtest environment as both the
+	// management and the workload cluster.
+	testEnvKubeconfig := kubeconfig.FromEnvTestConfig(testEnv.GetConfig(), &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "my-cluster",
+			Namespace: "my-namespace",
+		},
+	})
+	kubeconfigSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "my-cluster-kubeconfig",
+			Namespace: "my-namespace",
+		},
+		Data: map[string][]byte{
+			secret.KubeconfigDataName: testEnvKubeconfig,
+		},
+	}
+	clusterKey := client.ObjectKey{
+		Namespace: "my-namespace",
+		Name:      "my-cluster",
+	}
+
+	tests := []struct {
+		name       string
+		clusterKey client.ObjectKey
+		objs       []runtime.Object
+		expectErr  bool
+	}{
+		{
+			name:       "returns a workload cluster",
+			clusterKey: clusterKey,
+			objs:       []runtime.Object{etcdSecret.DeepCopy(), kubeconfigSecret.DeepCopy()},
+			expectErr:  false,
+		},
+		{
+			name:       "returns error if cannot get rest.Config from kubeconfigSecret",
+			clusterKey: clusterKey,
+			objs:       []runtime.Object{etcdSecret.DeepCopy()},
+			expectErr:  true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			for _, o := range tt.objs {
+				g.Expect(testEnv.Create(ctx, o)).To(Succeed())
+				defer cleanupResource(g, testEnv, o)
+			}
+
+			m := Management{
+				Client: testEnv,
+			}
+
+			workloadCluster, err := m.GetWorkloadCluster(context.Background(), clusterKey)
+			if tt.expectErr {
+				g.Expect(err).To(HaveOccurred())
+				g.Expect(workloadCluster).To(BeNil())
+				return
+			}
+			g.Expect(err).ToNot(HaveOccurred())
+			g.Expect(workloadCluster).ToNot(BeNil())
+		})
+	}
+
+}
+
+// cleanupResource removes the obj from the envtest environment. This should
+// be used to keep the envtest clean if the environment is being reused.
+func cleanupResource(g *WithT, testEnv *helpers.TestEnvironment, obj runtime.Object) {
+	var ss client.GracePeriodSeconds = 0
+	g.Expect(testEnv.Delete(context.Background(), obj, &ss)).To(Succeed(), "cleaning up")
```

Comment:
So it seems that this Delete will not actually remove the object from etcd. Ideally, I _think_ we should have an `Eventually` with a `testEnv.Get` and wait for us to get a `apiErrors.NotFound`. 
Maybe make some kind of generic helper that deletes the passed in `runtime.Object`. 🤷

Reply:
I'm leaning towards a different pattern, because I believe we should have each test (in this case table test) to run in a namespace with a random name, so:
- we ensure test isolation (every test start from a clean situation)
- cleaning up a test is always delete the namespace, no matter how many objects are in

I don't have yet figured out how to do so, but I think we should get some util in envtest like e.g.:
- create and delete NS
- a utility that given a list of objects, sets the namespace on all of them them
The tricky part is to get this utility to support also test case where we need some sort of control on namespaces (e.g. I need objects in two namespaces)